[For ALP][Don't merge now] Support corosync3/kronosnet

liangxin1300 opened the pull request:
# Note Before Merging
This PR is for corosync3, which is incompatible with corosync2. So, if the next step for the master branch is for 15SP6(which should still keep with corosync2), we should consider submitting this PR to another branch 
## Major changes:
- Support corosync3/kronosnet
- Totem UI change for node list and interface list
- Enhance validations for input
- Use 'black' to re-format Python file ag_corosync
- Bump version to 5.0.0
##  Screenshots
![Screenshot from 2023-07-06 11-35-50](https://github.com/yast/yast-cluster/assets/1650862/4f4d745e-6265-42bd-9c54-5807887455c0)
## Test rpm
- corosync3: https://build.opensuse.org/package/show/network:ha-clustering:Unstable/corosync
- knet: https://build.opensuse.org/package/show/network:ha-clustering:Unstable/kronosnet
  (remember all libknet1-* packages should be installed)
- yast2-cluster: https://build.opensuse.org/package/show/home:XinLiang:branches:openSUSE:Factory/yast2-cluster

nick-wang committed:
corosync3: Add new parameters and remove obsolete ones.

Add:
	system, resources, nozzle sub-directive
Remove:
	qb sub-directive
Change:
	timestamp default values

Modify read/write of totem to generic method.

nick-wang committed:
corosync3: Change interface and nodelist implementation.

Only support write 1 interface.
Only support add node section one by one.
Fix write of logger_subsys section.
Fix a bug with dict merge, should always go with .update()

Note: ag_openais ready

nick-wang committed:
corosync3: Move linknumber and knet parameters to interface.

Support multiple interfaces instead of only one.

nick-wang committed:
corosync3: Return IPs for doList nodelist.node

nick-wang committed:
corosync3: Fix return string wrapper should with " instead of '

nick-wang committed:
Only write logger_subsys/blackbox/function_name when necessary

nick-wang committed:
Recognize deprecated rrp_mode without write to check old format.

nick-wang committed:
corosync3: change ip_version and add fill iplist.

nick-wang committed:
corosync3: Totem UI change for node list and interface list.

nick-wang committed:
corosync3: Node list input table and remove corosync2 specific values.

nick-wang committed:
corosync3: support crypto model and change security dialog for knet.

nick-wang committed:
corosync3: Adapt helps

nick-wang committed:
corosync3: adapt udp ports

nick-wang committed:
corosync3: fix dev bugs.

1. Remove dummy write to logging section.

nick-wang committed:
corosync3: Remove bindnetaddr/mcastport/mcastaddr when udpu/knet

Set key to :interface_list Del, otherwise "l" will be used in cluster

nick-wang committed:
corosync3: Read the node list even with empty iplist.

Without a corosync.conf, yast will read /etc/corosync.conf.example
In the example file, node without IP configured in node list.

Should draw in UI instead of save in memory only.

nick-wang committed:
corosync3: polish the warning message and avoid duplicate node name.

nick-wang committed:
corosync3: maximum link number is 7 and smaller than configured rings

nick-wang committed:
corosync3: correct node name validation check

Need assign node name when multiple links instead of knet.

nick-wang committed:
corosync3: All nodes on the same link have the same IP family

nick-wang committed:
corosync3: fix the interface number and ringnum check.

nick-wang committed:
corosync3: Add link index in UI and change link number to selectbox.

nick-wang committed:
corosync3: fix broken interface input dialog

Fix broken UI when adding without existed linknumber.
Initialize knet transport.

nick-wang committed:
corosync3: allow discontinuous ringX_addr

nick-wang committed:
corosync3: disable/enable interface list based on selected transport

The interface sub-directive of totem is optional for UDP and knet transports.
For knet, multiple interface subsections define parameters for each knet
link on the system.
For UDPU an interface section is not needed and it is recommended that the
nodelist is used to define cluster nodes.

So, unicast should disable/drop interface list.
Multicast should drop knet_transport and knet_link_priority.
Kronosnet should drop bindnetaddr/mcastaddr/mcastport.

nick-wang committed:
corosync3: support check ipv4/ipv6

nick-wang committed:
corosync3: enhance based on demo feedback.

1. Change "Bind Address" to "Bind Net Addr", unified the labels
  of interface list table.
2. Immediate show warning when no Node ID of Kronosnet mode.
3. Enhance the comment of name in nodelist directive.

nick-wang committed:
corosync3: add validate rule of crypty cipher and move the switch value to correct place.

The checker is from corosync exec/totemconfig.c
    if ((strcmp(tmp_cipher, "none") != 0) &&
        (strcmp(tmp_hash, "none") == 0)) {
        *error_string = "crypto_cipher requires crypto_hash with
             value other than none";
        res = -1;

        goto out_free_crypto_model_str;
    }

nick-wang committed:
corosync3: do not save empty dialog to ip and interface list

1. Do not save empty input dialog to ip/interface list
2. Polish the warning message
3. IP label start from IP0 instead of IP1

nick-wang committed:
corosync3: totem Add cancel_hold_on_retransmit config option

In corosync's commit cdf72925d/64010f573, add
cancel_hold_on_retransmit config option to deliver packets
without corosync entering "FAILED TO RECEIVE" state

liangxin1300 committed:
Revert "Remove superfluous textdomain entry."

This reverts commit e75a7f094f78dd1aecbe07815a7a3eb1b2c2354a.

liangxin1300 committed:
Revert "Remove all openais stuff"

This reverts commit 0876c9ad75ece70a2bc3abe43b55d1660175a809.

liangxin1300 committed:
Remove all openais stuff

liangxin1300 committed:
Use 'black' to re-format python file ag_corosync

liangxin1300 committed:
Remove superfluous textdomain entry.

liangxin1300 committed:
Bump version to 4.6.1

liangxin1300 committed:
Minor changes

liangxin1300 committed:
Add more validations in nodelist_input_dialog

And remove unused codes

varkoly reviewed (commented):
Why have you reverted my openais changes which was already approved and merged 7 months ago?  https://github.com/yast/yast-cluster/pull/107

liangxin1300 commented:
> Why have you reverted my openais changes which was already approved and merged 7 months ago? #107

Because the python indent in https://github.com/yast/yast-cluster/blob/master/src/servers_non_y2/ag_corosync doesn't look good. That will bring more conflicts when cherry-pick Nick's corosync3 code.
So I revert it temporarily, after moved Nick's code, I rename it back:)

varkoly commented:
> 
Yes in original code there was a lot of tabs. I've removed these.

liangxin1300 committed:
Refactor the 'switch_*_button' functions by extracting the common parts of the code

liangxin1300 committed:
Adjust the security setting codes

liangxin1300 committed:
Don't configure 'Auto Generate Node ID' anymore

Corresponding no 'clear_node_high_bit' in corosync.conf
Due to the man corosync.conf, cluster behavior is undefined if this option enabled

liangxin1300 committed:
Change the validation function validate_nodeid and validate_name

liangxin1300 committed:
According to man corosync.conf, change ip_verion as 'ipv4' when transport is 'udp'

liangxin1300 committed:
Remove unused codes

- Remove 'Cluster.config_format'
- Remove 'memberaddr'

liangxin1300 committed:
Adjust linkmode codes

liangxin1300 committed:
Adjust interface input dialog codes

liangxin1300 committed:
Adjust qdevice setting codes

liangxin1300 committed:
Adjust codes for multicast and unicast configurations

liangxin1300 committed:
Update yast2-cluster.changes and bump version to 5.0.0

liangxin1300 commented:
# Note Before Merging


This PR is for corosync3, which is incompatible with corosync2. So, if the next step for the master branch is for 15SP6(which should still keep with corosync2), we should consider submitting this PR to another branch
